FIX #2486 Kubernetes provider - service - port was changed to ports

zeari opened the pull request:
port, container_port and protocol moved to a new model (container_service_port_config) to accommodate the changes to kubernetes listed here: [GoogleCloudPlatform/kubernetes/pull/6182](https://github.com/GoogleCloudPlatform/kubernetes/pull/6182)
This patch also migrates existing data.

zeari commented:
@abonas  done.

abonas commented on vmdb/app/models/container_service_port_config.rb:
```diff
@@ -0,0 +1,4 @@
+class ContainerServicePortConfig < ActiveRecord::Base
+  # :name :protocol :port :target_port
```
redundant comment...

abonas commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -230,6 +233,16 @@ def parse_container_port_config(port_config, pod_id, container_name)
 
+    def parse_service_port_config(port_config, service_id)
+      {
+        :ems_ref     => "#{service_id}_#{port_config["port"]}_#{port_config["targetPort"]}_#{port_config["protocol"]}",
```
why is protocol needed as part of ems_ref?

simon3z replied:
Maybe in this case `ems_ref` can be omitted altogether:

* service ports are not an entity on their own (they are just a one-to-many items of services)
* using **all** the attributes as `ems_ref` basically means that there's no field that can be updated without generating a new entity

If `save_inventory_multi` supports it I'd suggest to remove the `ems_ref` in this specific case. (/cc @Fryguy )

abonas replied:
> Maybe in this case ems_ref can be omitted altogether:
> 
>     service ports are not an entity on their own (they are just a one-to-many items of services)
>     using all the attributes as ems_ref basically means that there's no field that can be updated without generating a new entity
> 
> If save_inventory_multi supports it I'd suggest to remove the ems_ref in this specific case. (/cc @Fryguy )

this is similar to what was done in containers/containers def/container port configs - they all don't have a "real" ems_ref. iirc, it was advised by @Fryguy to have those ems_ref fields generated (@Fryguy pls correct me if I'm wrong).
I'm OK to drop them if they are not needed. And if we're gonna drop them -imo  it has to be done for all the "nested" entities.

simon3z replied:
@abonas yes but at least in containers you have some fields that can be updated (`restart_count`, `state`, etc.). Here it's pretty much nothing at this point.

Fryguy replied:
You need some sort of unique identifier otherwise the save inventory code
will thrash on every refresh (every record will be an add and a delete).
You can either use ems_ref or have a multi-value key in the save inventory.

ems_ref is technically supposed to be the actual value for an entity from
the EMS, so you can communicate that value back to the native API if
needed.  However we've built keys like this before if it didn't make sense,
or a real value isn't available.

The choice for having ems_ref versus a multi key is also driven by multiple
providers.   In this case, what is the equivalent for Amazon containers,
and does that equivalent have a unique identifier.  If so, then you have to
have ems_ref; if not then either way is fine.

/cc @blomquisg
 On Apr 7, 2015 7:13 AM, "abonas" <notifications@github.com> wrote:

> In vmdb/app/models/ems_refresh/parsers/kubernetes.rb
> <https://github.com/ManageIQ/manageiq/pull/2523#discussion_r27868681>:
>
> > @@ -230,6 +233,16 @@ def parse_container_port_config(port_config, pod_id, container_name)
> >        }
> >      end
> >
> > +    def parse_service_port_config(port_config, service_id)
> > +      {
> > +        :ems_ref     => "#{service_id}_#{port_config["port"]}_#{port_config["targetPort"]}_#{port_config["protocol"]}",
>
>  Maybe in this case ems_ref can be omitted altogether:
>
> service ports are not an entity on their own (they are just a one-to-many items of services)
> using all the attributes as ems_ref basically means that there's no field that can be updated without generating a new entity
>
> If save_inventory_multi supports it I'd suggest to remove the ems_ref in
> this specific case. (/cc @Fryguy <https://github.com/Fryguy> )
>
> this is similar to what was done in containers/containers def/container
> port configs - they all don't have a "real" ems_ref. iirc, it was advised
> by @Fryguy <https://github.com/Fryguy> to have those ems_ref fields
> generated (@Fryguy <https://github.com/Fryguy> pls correct me if I'm
> wrong).
> I'm OK to drop them if they are not needed. And if we're gonna drop them
> -imo it has to be done for all the "nested" entities.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ManageIQ/manageiq/pull/2523/files#r27868681>.
>

abonas commented:
This patch should not be merged until openshift is rebased again on k8s.

@zeari , according to @simon3z endpoints keep changing (together with
services). So this patch will have to deal with that as well (parse
endpoints) in order to keep the services => endpoints relationship.
Roughly the opposite of:
 https://github.com/simon3z/manageiq/commit/8d8a0b2

simon3z commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -133,14 +136,24 @@ def parse_endpoint(entity)
-          endpoint.targetRef["table"][:namespace], endpoint.targetRef.name)
-        new_result[:container_groups] << cg unless cg.nil?
+      # TODO: remove once kubernetes api v1beta3/v1.0 will be stable
+      if !entity.respond_to?(:subsets) || entity.subsets.nil?
```
You can remove the comment and the entire block. We know that we moved to the newer kubernetes.

simon3z commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -133,14 +136,24 @@ def parse_endpoint(entity)
+        return new_result
+      end
+
+      entity.subsets.each do |subset|
```
this now becomes:

    (entity.subsets || []).each do |subset|

zeari replied:
@simon3z done.

simon3z commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -133,14 +136,16 @@ def parse_endpoint(entity)
-          # a non method access
-          endpoint.targetRef["table"][:namespace], endpoint.targetRef.name)
-        new_result[:container_groups] << cg unless cg.nil?
+      entity.subsets.each do |subset|
```
I think you missed:

    (entity.subsets || []).each do |subset|

zeari commented:
@abonas Please review.

Fryguy commented on vmdb/app/helpers/container_service_helper/textual_summary.rb:
```diff
@@ -7,16 +7,18 @@ def textual_group_properties
   end
 
+  def textual_port_configs
+    items = ContainerServicePortConfig.where( :container_service_id => @record.id)
```
Minor, but no space after the `.where(`

Fryguy commented on vmdb/app/helpers/container_service_helper/textual_summary.rb:
```diff
@@ -53,12 +52,16 @@ def textual_portal_ip
+  def textual_port_config(single_conf)
+    name = single_conf.name
+
+    name = "Unnamed" if name.nil? || name.empty?
```
This can be simplified to `name = "Unnamed" if name.blank?`

@h-kataria @dclarizio Do you have some common string you use for unnamed things to be consistent elsewhere in the UI?

dclarizio replied:
We normally use something like "<Unnamed>", just in case someone actually calls something "Unnamed".
This should also use gettext, i.e. _("<Unnamed>").
@mfeifer Can you suggest something instead of "<Unnamed>"?  Maybe "<Name not provided>" or something?

Fryguy commented on vmdb/app/helpers/container_service_helper/textual_summary.rb:
```diff
@@ -53,12 +52,16 @@ def textual_portal_ip
+
+    {
+      :label => name,
+      :value => "#{single_conf.protocol} port:#{single_conf.port} to port:#{single_conf.target_port}"
```
Is this a port range?  It looks like one, but the presentation seems odd.  If so, @h-kataria @dclarizio Is there a common way to present port ranges?

zeari replied:
@Fryguy Its not a port range. 
This string represents a single [service port](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/types.go#L957) from a service.

abonas replied:
@zeari , maybe it will be more clear if the second part will be "to target port X" instead of "to port"
it will also resonate better with the original field name.

zeari replied:
@abonas string ports look a little funny this way
![new format](https://cloud.githubusercontent.com/assets/11256940/7180172/cd8f8fd8-e447-11e4-8648-98c883f49c2b.png)

abonas replied:
perhaps target_port? it's not ideal, but that's the field name, and it will look less like a port range.
or maybe add some verb there next to the "to"? 
worth to check the target port definition on k8s side, maybe there's something there that makes sense to use here?

zeari replied:
```go
// Optional: The target port on pods selected by this service.
// If this is a string, it will be looked up as a named port in the
// target Pod's container ports. If this is not specified, the value
// of Port is used (an identity map).
 TargetPort util.IntOrString `json:"targetPort" description:"the port to access on the pods targeted by the service; defaults to the service port"
```
How about this:
![port_on_pod](https://cloud.githubusercontent.com/assets/11256940/7180521/05ab78ca-e44b-11e4-9259-a3b999f14d12.png)

abonas replied:
@zeari :+1:

Fryguy commented on vmdb/app/helpers/container_service_helper/textual_summary.rb:
```diff
@@ -7,16 +7,18 @@ def textual_group_properties
     items.collect { |m| send("textual_#{m}") }.flatten.compact
   end
 
+  def textual_port_configs
```
Is this a new group in the summary page?  If so, this should be named `textual_group_port_configs` and needs exposure from the view, which I don't see in this PR, so will be needed.

Can you attach a screenshot showing these changes?

zeari replied:
@Fryguy 
![gui-example](https://cloud.githubusercontent.com/assets/11256940/7180008/109cc072-e446-11e4-8c82-c24a58bb2b3f.png)
[Here](https://github.com/ManageIQ/manageiq/pull/2523/files#r28372557) is where this method is called and port configurations are displayed.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -216,6 +221,16 @@ def parse_container_port_config(port_config, pod_id, container_name)
+        :name        => port_config["name"],
+        :port        => port_config["port"],
+        :target_port => port_config["targetPort"],
+        :protocol    => port_config["protocol"]
```
I thought dot notation was the way to do this now.  cc @abonas

abonas replied:
> I thought dot notation was the way to do this now. cc @abonas

typically yes, but in some cases it doesn't work due to recursiveopenstruct design.
actually, @zeari, since kubeclient was updated to its latest version, it's also configured to support dot notation when array is used as part of the object, not only hash. so worth to try now.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -133,14 +136,16 @@ def parse_endpoint(entity)
-        new_result[:container_groups] << cg unless cg.nil?
+      (entity.subsets || []).each do |subset|
+        (subset.addresses || []).each do |address|
+          next if address.targetRef.nil? || address.targetRef.kind != 'Pod'
```
I could have sworn I commented on this already (@simon3z, right?), but this can simplify to `next if address.targetReg.try(:kind) == 'Pod'`

simon3z replied:
I suppose it should be (very similar to what you suggest, except for the logic):

    next if address.targetReg.try(:kind) != 'Pod'

Fryguy commented on vmdb/app/views/container_service/_main.html.haml:
```diff
@@ -2,5 +2,6 @@
 .row
   .col-sm-12.col-md-12.col-lg-6
     = render :partial => "shared/summary/textual", :locals => { :title => _("Properties"), :items => textual_group_properties }
+    = render :partial => "shared/summary/textual", :locals => { :title => _("Port Configurations"), :items => textual_port_configs }
```
LOL I'm blind.  Even so, the naming is inconsistent.

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def copy_over_configs
+    ContainerService.find_each do |service|
```
This should just be `.all.each` instead of `find_each`

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,54 @@
+
+  def copy_over_configs
+    ContainerService.find_each do |service|
+      ContainerServicePortConfig.create(
```
Prefer `create!` over `create`

Fryguy replied:
- Avoid using models directly in migrations.  You will have to use stubs for data access.
- Migration specs are required for data migrations.
- When you have a section with a data migration wrap it in a `say_with_time` block.  You can see that in other data migrations, and it's used to show progress and time to complete in the migration output.

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,54 @@
+  def copy_over_configs
+    ContainerService.find_each do |service|
+      ContainerServicePortConfig.create(
+          :ems_ref              => "#{service[:ems_ref]}_#{service[:port]}_"\
```
Please stop using the line continuation character (anywhere).  It's not necessary.  Two strings on two separate lines become one string automatically.  There is almost never a case you need it in Ruby

cc @abonas @simon3z

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,54 @@
+  def down
+    add_columns
+
+    ContainerServicePortConfig.find_each do |port_config|
```
Prefer `.all.each` over `.find_each`

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def down
+    add_columns
```
I'm finding the extraction into method makes these migrations more complex.  Can we undo that?  Its pretty uncommon to do that in migrations.

zeari replied:
@Fryguy Rubocop complained the methods were too long so I split them up into smaller methods.
cc @abonas

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,54 @@
+    add_columns
+
+    ContainerServicePortConfig.find_each do |port_config|
+      s = ContainerService.where(:id => port_config[:container_service_id]).first # There should be only one
```
Doing this query in a loop can be expensive.  You can do this outside the loop once and use `index_by` to put it into a Hash: `indexed_containers = ContainerService.all.index_by(&:id)`.  Then that Hash can be used in the loop withioutt the N+1 penalty: `s = indexed_containers[port_config[:container_service_id]]`

Fryguy replied:
Additionally, since port_config is an ActiveRecord object, prefer dot access to attributes over `[]` access.

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,54 @@
+      s = ContainerService.where(:id => port_config[:container_service_id]).first # There should be only one
+      s[:port] = port_config[:port]
+      s[:protocol] = port_config[:protocol]
+      s[:container_port] = port_config[:target_port]
```
Also use `.method=` access to set attributes over `[]=` access

zeari commented:
@abonas @Fryguy 
Changed the code according to comments
* Added spec file for this migration
* I didnt fix the AbcSize comments from Rubocop because of:

> I'm finding the extraction into method makes these migrations more complex. Can we undo that?...

* Added migration stubs instead of using models directly
* Dot access to attributes over [] access.

simon3z commented on vmdb/app/models/container_service.rb:
```diff
@@ -6,6 +6,7 @@ class ContainerService < ActiveRecord::Base
 
   belongs_to  :ext_management_system, :foreign_key => "ems_id"
   has_and_belongs_to_many :container_groups
+  has_many :container_service_port_configs
```
I think you're maintaining the `configs` suffix here for consistency with `container_port_configs`. But actually it seems redundant to specify that it's a configuration. I think that the naming of these tables should be `container_ports` and `container_service_ports`.

zeari replied:
@simon3z The problem with leaving it without a 'config' suffix is that its a little confusing. 
When I'm looking at a "service_port" field, what Im thinking is "This is a 4 digit port integer"
When I'm looking at "port_config" Im thinking about a data structure with multiple fields (which it is)

abonas replied:
:dependent => :destroy

zeari commented:
@Fryguy Can this be merged?

abonas commented on vmdb/app/helpers/container_service_helper/textual_summary.rb:
```diff
@@ -53,12 +52,16 @@ def textual_portal_ip
 
-  def textual_container_port
-    {:label => "Container Port", :value => @record.container_port}
+  def textual_port_config(single_conf)
```
consider changing single_conf to a more meaningful name

abonas commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,56 @@
+    say_with_time("Moving container service port records to container service port config table") do
+      ContainerService.all.each do |service|
+        ContainerServicePortConfig.create!(
+            :ems_ref              => "#{service.ems_ref}_#{service.port}_#{service.container_port}_#{service.protocol}",
```
what about the protocol field? still part of ems_ref? iirc, there was a discussion to remove it from ems_ref...

zeari replied:
> what about the protocol field? still part of ems_ref? iirc, there was a discussion to remove it from ems_ref...

@abonas Your'e right. It was removed from the parser but not from here.

abonas commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,56 @@
+    say_with_time("Moving container service port config records back to container service") do
+      indexed_containers = ContainerService.all.index_by(&:id)
+
+      ContainerServicePortConfig.all.each do |port_config|
```
Is "each" needed here or .first might be suffice? (given that you can only fit one port in the original design when it's part of the services table)

abonas commented on vmdb/spec/migrations/20150405141637_remove_port_config_from_container_service_spec.rb:
```diff
@@ -0,0 +1,39 @@
+      pconfig = container_service_port_config_stub.where(:container_service_id => service.id).first
+      pconfig.protocol.should    == "TCP"
+      pconfig.port.should        == 1111
+      pconfig.target_port.should == "2222" # container_port:integer turns into target_port:string
```
Why does it turn into a string?

zeari replied:
> Why does it turn into a string?

@abonas  because target_port can be 'redis-server'

jrafanie replied:
Sorry, i'm naive here, why is `target_port` ever `redis-server`  ???  It's strange to have numeric ports except when it's `redis-server`.

zeari replied:
> Sorry, i'm naive here, why is target_port ever redis-server ??? It's strange to have numeric ports except when it's redis-server.

@jrafanie In hindsight, that wasn't a good answer. This code is out of the Kubernetes [types.go](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/types.go#L1002) file. 
```go
// Optional: The target port on pods selected by this service.  If this
// is a string, it will be looked up as a named port in the target
// Pod's container ports.  If this is not specified, the first port on
// the destination pod will be used.  This behavior is deprecated.  As
// of v1beta3 the default value is the sames as the Port field (an
// identity map).
TargetPort util.IntOrString `json:"targetPort"`
```
Basically, ports can be named instead of just four digit numbers. 'redis-server' is just an example.

jrafanie replied:
@zeari So, we don't have a choice, right?  We could get a either a numeric or a string "service name"... My concern is that makes things more complicated.  In an ideal world, we'd get all numerics in the database and provide "user friendly" lookups of the port to "service name".  Regarding complexity, imagine trying to run a report of `target_port` where we can no longer sort by that field since it's now a complex type.

If we have no choice, that's all we can do.

zeari replied:
> If we have no choice, that's all we can do.

@abonas @jrafanie Pretty much. That is how Kubernetes is designed (ATM)
We can always sort alphanumerically if we have to.

jrafanie replied:
Ok, thanks @zeari.  :cry:

abonas commented on vmdb/spec/models/ems_refresh/refreshers/k8s_refresher_spec.rb:
```diff
@@ -11,13 +11,14 @@
     end
     @ems.reload
 
+    # All ems_ref fields and other auto generated fields arn't checked because the VCR file needs update
```
s/arn't/aren't

jrafanie commented:
@abonas Does this look good to you?  I'm not familiar with the change.

abonas commented on vmdb/app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -108,16 +108,19 @@ def parse_service(service)
       )
+
+      ports = service.spec.ports
+      new_result[:container_service_port_configs] = Array(ports).collect do |port_entry|
```
@zeari, please remind me why is there a need here to do Array(ports)?

abonas commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,56 @@
+    say_with_time("Moving container service port config records back to container service") do
+      indexed_services = ContainerService.all.index_by(&:id)
+
+      ContainerServicePortConfig.all.each do |port_config|
```
is all.each necessary? you might have here multiple ports, but the services table has only one record that can hold one port...

zeari replied:
> is all.each necessary? you might have here multiple ports, but the services table has only one record that can hold one port...

@abonas  I guess changing it to take the first ```port_config``` using ```service.container_service_port_configs.first``` for each service is more efficient.

abonas replied:
> @abonas I guess changing it to take the first port_config using service.container_service_port_configs.first for each service is more efficient.

Was this tested with more than one port in ports section?

zeari replied:
> Was this tested with more than one port in ports section?

@abonas yes, it just takes the last ```port_config ``` in the tables for each service.

abonas commented:
> @abonas Does this look good to you? I'm not familiar with the change.

I'm familiar with the change and it LGTM. I left a minor comment/question for @zeari , lets wait for his response before merging.

zeari commented:
> I'm familiar with the change and it LGTM. I left a minor comment/question for @zeari , lets wait for his response before merging.

@abonas Changed and pushed

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,55 @@
+            :target_port          => service.container_port,
+            :container_service_id => service.id
+        )
+        say "service '#{service.name}' updated!"
```
We generally don't emit this much detail into the migrations, so this should be removed.

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,55 @@
+    add_column :container_services, :port, :integer
+    add_column :container_services, :protocol, :string
+    add_column :container_services, :container_port, :integer
+    ContainerService.reset_column_information
```
This line should not be needed.

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,55 @@
+class RemovePortConfigFromContainerService < ActiveRecord::Migration
+  class ContainerService < ActiveRecord::Base
+    has_many :container_service_port_configs
```
This has_many reference needs to use the full namespace of the stub.

```ruby
has_many :container_service_port_configs, :class_name => "RemovePortConfigFromContainerService::ContainerServicePortConfig"
```

Fryguy commented on vmdb/db/migrate/20150405141637_remove_port_config_from_container_service.rb:
```diff
@@ -0,0 +1,55 @@
+        service.update_attributes!(:port           => port_config.port,
+                                   :protocol       => port_config.protocol,
+                                   :container_port => port_config.target_port)
+        say "service '#{service.name}' updated!"
```
Also remove this `say`.

Fryguy commented:
Looks great @zeari .  Most of my comment are relatively simple fixes, so once those are in we can merge.

Fryguy commented on vmdb/spec/models/ems_refresh/refreshers/k8s_refresher_spec.rb:
```diff
@@ -39,42 +40,41 @@ def assert_ems
     @container = Container.find_by_name("heapster")
     @container.should have_attributes(
-      :ems_ref       => "49984e80-e1b7-11e4-b7dc-001a4a5f4a02_heapster_kubernetes/heapster:v0.9",
+      # :ems_ref       => "a7566742-e73f-11e4-b613-001a4a5f4a02_heapster_kubernetes/heapster:v0.9",
```
@blomquisg How did you solve the problem of the ids constantly changing for the other providers?  I can't remember.

simon3z commented:
@zeari @abonas @Fryguy at the moment this patch is high priority because it's holding us back from fully testing newer kubernetes. Can we handle it ASAP (I think there are still few comments that @zeari needs to address but the rest looks fine).

zeari commented:
@simon3z Ill get to this as soon as I can.

zeari commented:
@simon3z @Fryguy @abonas 
Updated and rebased. Theres one final problem I cant figure out in the migration spec.
If the test case for ```:down``` runs before ```:up```  the test succeeds. 
If its the other way around I get errors from the spec complaining about db relationships. 
[full error log](http://pastebin.com/2rAxnnxE)

Fryguy commented:
@zeari I missed you comment about the test stuff...let me take a look
